### DISCUSSION TITLE
Multiple dataloaders for val/test in the form of Dict in LightningDataModule

### DISCUSSION THREAD

**Discussion by woozch** [Lightning-AI/pytorch-lightning#12048] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
Hi.

I am currently developing a multi-task model that evaluates some datasets with different evaluation metrics for each, but it is impossible to configure multiple val/test dataloader in the form of a dictionary(mapping).

https://github.com/PyTorchLightning/pytorch-lightning/blob/2d043857eef55eb327528b70c8b33c680e82fb7b/pytorch_lightning/core/datamodule.py#L232

It seems that configuring the dataloader in the form of Dict in LightningDataModule is only possible with train_dataloader.
Why not make it possible in val_dataloader or test_dataloader? Is there any other reason for this design?

**Comment by rohitgr7** [comment]
hey @wgchang !

within `train_dataloader` when a dict is returned, it creates a CombinedLoader, instead of a sequence of dataloaders.
check out this thread here: https://github.com/PyTorchLightning/pytorch-lightning/discussions/5782#discussioncomment-2121730

**Reply by woozch** [reply]
Thanks for replying.
In the case of train_dataloader, I confirmed that each DataLoader is returned in the form of a dictionary, but in the case of val_dataloader and test_dataloader, it did not work, so I asked a question.

```python
class MyDataModule(DataModule):
    ....
    def val_dataloader(self):
        # working
        return [Dataloader(...), Dataloader(...)] # okay
        # not working
        return { 'validation_set1': Dataloader(...),
                 'validation_set2': Dataloader(...) }. # it seems not allowed
 ```
If each validation set is evaluated with a different performance metric, I would like to evaluate each validation set with the corresponding metric. For example, accuracy on validation_set1 and MSE on validation_set2.

**Reply by rohitgr7** [reply]
yep, it's not allowed. It depends how you want to iterate over the multiple dataloaders. Returning a dict is not allowed.
supported configurations for `val_dataloaders` is mentioned here: https://pytorch-lightning.readthedocs.io/en/latest/guides/data.html#multiple-validation-test-predict-dataloaders

**Reply by rohitgr7** [reply]
for your use case, I'd suggest using as a list of dataloaders where you get a `dataloader_idx` inside your `validation_step` and you can make a distinction there to configure which metric to use for which dataloader.

**Reply by woozch** [reply]
Thank you. I'll try to write the code in the way you suggested.

**Reply by thesofakillers** [reply]
> yep, it's not allowed

you should make it allowed.

**Reply by surya-narayanan** [reply]
I agree, why is there a distinction between the expected behavior of training and test steps?
